Fix part of #8423: Add lint check to ensure there's a space before the opening braces of a function start

Mayank-gaur opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #8423 .
2. This PR does the following: [Add lint check to ensure there's a space before the opening braces of a function start]

## Essential Checklist

- [yes ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ yes] The linter/Karma presubmit checks have passed locally on your machine.
- [ yes] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [ yes] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

![Screenshot from 2021-04-03 04-27-34](https://user-images.githubusercontent.com/56195849/113460102-31866180-9435-11eb-92ab-8e4dd79cbd1e.png)
All the test cases that I have written are successfully passing. Also, pre-commit linter was also passing successfully.
If I alter a file by removing space before the opening brace, error is also thrown as shown below,

![Screenshot from 2021-04-03 22-03-09](https://user-images.githubusercontent.com/56195849/113484951-b4a4c780-94c8-11eb-8660-fdc2a4392d58.png)


@Hudda, @DubeySandeep, @sajalasati  PTAL.

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Mayank-gaur committed:
Added a custom eslint extension and corresponding tests.

Mayank-gaur committed:
Fixed linting issues

Mayank-gaur committed:
Fixed linting issues

Mayank-gaur committed:
Fixed linting issues

Hudda commented on scripts/linters/custom_eslint_checks/rules/space-before-opening-brace.js:
```diff
@@ -0,0 +1,56 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
2021*

Mayank-gaur replied:
Corrected!

Hudda commented on scripts/linters/custom_eslint_checks/rules/space-before-opening-brace.js:
```diff
@@ -0,0 +1,56 @@
+    type: 'layout',
+    docs: {
+      description:
+        'disallow more than one brace before opening brace of function',
```
```suggestion
        'Enforce a single space before opening braces of a function start',
```

Mayank-gaur replied:
Done!

Hudda commented on scripts/linters/custom_eslint_checks/rules/space-before-opening-brace.js:
```diff
@@ -0,0 +1,56 @@
+    return {
+      FunctionDeclaration: function(node) {
+        const lines = sourcecode.getText(node);
+        for (var i = 0; i < lines.length; i++) {
+          if (lines[i] === ')') {
+            if (lines[i + 1] !== ' ' || lines[i + 2] !== '{') {
```
Instead of parsing the full line, check only two characters before `{`.

Mayank-gaur replied:
Thanks @Hudda Done. I have used indexOf() function to search for the braces. Linter is working fine also. I earlier forgot to add a screenshot that test is failing where it need to, I have added that too in the PR description. .

Hudda commented on scripts/linters/custom_eslint_checks/rules/space-before-opening-brace.js:
```diff
@@ -0,0 +1,56 @@
+    fixable: null,
+    schema: [],
+    messages: {
+      OnlyOneSpacebeforebrace:
```
```suggestion
      spaceBeforeBrace:
```

Mayank-gaur replied:
Corrected.

Hudda commented on scripts/linters/custom_eslint_checks/rules/space-before-opening-brace.spec.js:
```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
2021*

Mayank-gaur replied:
Corrected.

Hudda reviewed (changes requested):
Thanks, @Mayank-gaur I've left some comments PTAL!

Mayank-gaur committed:
1) Corrected dates to 2021.
2) searched only for braces and its value before them.
3) Changed meta desc. and msgId.

Mayank-gaur commented:
> Thanks, @Mayank-gaur I've left some comments PTAL!

Thanks for guiding me. I have done the requested changes. @Hudda PTAL.

Hudda reviewed (approved):
LGTM, Thanks @Mayank-gaur!

Hudda commented:
@DubeySandeep @vojtechjelinek PTAL!

DubeySandeep commented on scripts/linters/custom_eslint_checks/rules/space-before-opening-brace.js:
```diff
@@ -0,0 +1,54 @@
+// Copyright 2021 The Oppia Authors. All Rights Reserved.
```
@Mayank-gaur Eslint already provides a check [space-before-blocks](https://eslint.org/docs/rules/space-before-blocks) but because of the typescript parser we the checks were not working correctly. We had reported this to eslint and had a tracking issue at #10470. It looks like the issue is resolved and once we will update our eslint version this check will start working!

@Hudda, Can you please confirm^? Also, do you think we still need the check introduced in this PR?

/cc @sajalasati

Hudda replied:
I think we should make sure if updating the eslint version fixes the issue before closing this PR.

DubeySandeep replied:
I tried testing the existing feature in our existing eslint version (7.1.0) and I can't reproduce any issue:
![image](https://user-images.githubusercontent.com/16653571/113968685-31c99700-9851-11eb-9f4a-58010cf6d3bb.png)


@Hudda Can you please verify? and if that's true then let's close #10470!

Mayank-gaur commented:
@DubeySandeep If a check is already available for this, I dont think that we need to add another custom extension. @Hudda , @sajalasati and @DubeySandeep should I close this PR? Please guide me.

sajalasati commented:
@Mayank-gaur Once we have resolved #10470 and verified that the built-in check works as expected, we can safely close this PR.

Also, please don't use the option to request review. It works in a bit weird manner, and leaves reviewer in confusion. Always use PTAL comment to assign a person, if you can't do it directly.

Mayank-gaur commented:
> @Mayank-gaur Once we have resolved #10470 and verified that the built-in check works as expected, we can safely close this PR.
> 
Thanks  for the update. I wanted to ask that can I work on other tasks untill #10470 is resolved?
> Also, please don't use the option to request review. It works in a bit weird manner, and leaves reviewer in confusion. Always use PTAL comment to assign a person, if you can't do it directly.

@sajalasati  Sorry for the trouble.

sajalasati commented:
> Thanks  for the update. I wanted to ask that can I work on other tasks untill #10470 is resolved?

Yes! And I think you already have other tasks assigned, active PRs right?

Mayank-gaur commented:
> > Thanks  for the update. I wanted to ask that can I work on other tasks untill #10470 is resolved?
> 
> Yes! And I think you already have other tasks assigned, active PRs right?

Yes @sajalasati I am assigned to other tasks,  but can you PTAL  [here](https://github.com/oppia/oppia/issues/8423#issuecomment-814278254), as I am not able to have any progress by myself in this task. Thanks!

sajalasati commented:
@Hudda Should we update the PR title to denote this is blocked on another issue #10470? The next action on this PR has to be taken by you. Thanks!

Mayank-gaur commented:
Please confirm what should I do with this PR? @Hudda PTAL.

Hudda commented:
Closing this PR as updating eslint fixes the issue.

sajalasati commented:
@Hudda Can you please also update the status of this task in #8423? (Strangely I wasn't able to locate this there)

Hudda commented:
> 
> 
> @Hudda Can you please also update the status of this task in #8423? (Strangely I wasn't able to locate this there)

Same, I didn't find any task with this name.
